include/hlslib/common/OpenCL.h

```diff
@@ -620,6 +714,113 @@ class Buffer {
     CopyToDevice(0, nElements_, other, 0);
   }
 
+/*
+The following functions copy 3D blocks of memory between host and device or on the device.
+xxxOffset -> the offset of destination or target in (elements, rows, slices).
+             For 2d arrays xxxOffset[2] should be 0.
+copyBlockSize -> The size of the block that is actually copied in (elements, rows, slices).
+hostBlockSize/deviceBlockSize/sourceBlockSize/destBlockSize: The size of the whole host/device/
+             source/destination Array in (elements, rows, slices)
+
+Note: (elements, rows, slices) means just "default"-indices, like one would index a default 
+c++ multidimensional array. 
+
+For hostBlockSize/deviceBlockSize/sourceBlockSize/destBlockSize it should hold that their product is
+smaller or equal to the size of the array they reference. Not explicitely checked, because openCL will
+throw an error if violated.
+*/
+template <typename IteratorType, typename = typename std::enable_if<
+                                      IsIteratorOfType<IteratorType, T>() &&
+                                      IsRandomAccess<IteratorType>()>::type>
+  void CopyBlockFromHost(const std::array<size_t, 3> hostBlockOffset, const std::array<size_t, 3> deviceBlockOffset,
+                        const std::array<size_t, 3> copyBlockSize, const std::array<size_t, 3> hostBlockSize,
+                        const std::array<size_t, 3> deviceBlockSize, IteratorType source) {
```

Comment:
Using an `std::array` is quite specific, how about accepting anything that we can index into?

Reply:
For some inspiration: https://devblogs.microsoft.com/oldnewthing/20190619-00/?p=102599